### PR TITLE
Add `--warnings warn` flag to `x.py`

### LINKED ISSUE #76805: bootstrap: `--warnings warn` doesn't apply to bootstrap itself
Bootstrapping with `RUSTFLAGS="-Wsomething that warns ./x.py check  --warnings warn ` does not work; a warning inside bootstrap will still abort the compilation process:

````
error: trivial numeric cast: `u32` as `u32`
   --> src/bootstrap/lib.rs:156:31
    |
156 |             libc::setpriority(libc::PRIO_PGRP as _, 0, 10);
    |                               ^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D trivial-numeric-casts` implied by `-D warnings`
    = help: cast can be replaced by coercion; this might require a temporary variable

error: aborting due to previous error

error: could not compile `bootstrap`.

````

EDIT: well, putting `--cap-lints warn` into `RUSTFLAGS` works but it seems very unintuitive.

### PR DESCRIPTION
**Pull request by Kobzol**
So that bootstrap itself can be built with warnings not being treated as errors.

Fixes: https://github.com/rust-lang/rust/issues/76805

r? @jyn514

### CODE REVIEW

File: src/bootstrap/bootstrap.py [hunk 1]
```diff
@@ -977,6 +977,7 @@ def parse_args():
     parser.add_argument('--color', choices=['always', 'never', 'auto'])
     parser.add_argument('--clean', action='store_true')
     parser.add_argument('--json-output', action='store_true')
+    parser.add_argument('--warnings', choices=['deny', 'warn', 'default'])
```
Comment (jyn514) [inline src/bootstrap/bootstrap.py:980]: ah hmm, I just realized bootstrap itself allows `--warnings default` even though it's not documented - seems ok to allow it here for consistency.

File: src/bootstrap/bootstrap.py [hunk 2]
```diff
@@ -888,7 +888,7 @@ def build_bootstrap(self, color, verbose_count):
         if target_linker is not None:
             env["RUSTFLAGS"] += " -C linker=" + target_linker
         env["RUSTFLAGS"] += " -Wrust_2018_idioms -Wunused_lifetimes"
-        if self.get_toml("deny-warnings", "rust") != "false":
+        if self.get_toml("deny-warnings", "rust") != "false" and warnings != "warn":
```
Comment (jyn514) [inline src/bootstrap/bootstrap.py:891]: this doesn't look right - it means `--warnings deny` will be ignored if you have `rust.deny-warnings=false` in config.toml. I think the logic should be something like
```suggestion
        if warnings == "default":
            deny_warnings = self.get_toml("deny-warnings", "rust") != "false"
        else:
            deny_warnings = warnings == "deny"
        if  deny_warnings:
```

Comment (Kobzol) [reply src/bootstrap/bootstrap.py:891]: Ah, I didn't think of precedence. Thanks.

File: src/bootstrap/bootstrap.py [hunk 3]
```diff
@@ -888,7 +888,11 @@ def build_bootstrap(self, color, verbose_count):
         if target_linker is not None:
             env["RUSTFLAGS"] += " -C linker=" + target_linker
         env["RUSTFLAGS"] += " -Wrust_2018_idioms -Wunused_lifetimes"
-        if self.get_toml("deny-warnings", "rust") != "false":
+        if warnings == "default":
```
Comment (jyn514) [inline src/bootstrap/bootstrap.py:891]: I gave you the wrong suggestion code, sorry, this is unset by default
```suggestion
        if warnings is None or warnings == "default":
```
(or add `default="default"` in the argparse code below)

**Review by jyn514 — CHANGES_REQUESTED** [review]
r=me with the bug fixed

File: src/bootstrap/bootstrap.py (hunk 3 above)
Comment (Kobzol) [reply src/bootstrap/bootstrap.py:891]: Uhh, argparse is a piece of crap :D I always use `click` or `typer`, so I forgot about this.

**Comment by Kobzol** [comment]
<at>bors r=jyn514

(I don't have `r` rights :) ).

**Comment by jyn514** [comment]
ah, we should fix that - I'll see if I can talk to t-infra about inviting you :)

oh, I just remembered this should probably have a test - I don't think we have an easy test suite for the arguments we pass to cargo in bootstrap though, so I'll add it at the same time I fix https://github.com/rust-lang/rust/pull/111979#issuecomment-1568533620.

@bors r+ rollup

**Pull request merged**